src/prefect/task_engine.py

```diff
@@ -591,7 +591,8 @@ def setup_run_context(self, client: Optional[SyncPrefectClient] = None):
                     log_prints=log_prints,
                     task_run=self.task_run,
                     parameters=self.parameters,
-                    result_factory=run_coro_as_sync(ResultFactory.from_task(self.task)),  # type: ignore
+                    # TODO: Allow reuse of result factories
+                    result_factory=ResultFactory(),  # type: ignore
```

Comment:
so my proposal is kind of a hybrid of the previous `classmethod` loader-approach and your new _fully_ encapsulated approach.  I think the result factory class should be "dumb" and take what it's given.

The edge case is populating a class without all the necessary information - I think this can be handled by requiring that information at _method call time_; so `ResultFactory().create(obj)` would error out with "the following kwargs need to be provided: serializer, etc."